openvdb_points/tools/PointGroup.h

```diff
@@ -248,6 +259,59 @@ struct SetGroupFromIndexOp
 }; // struct SetGroupFromIndexOp
 
 
+template <typename PointDataTree, typename FilterT>
+struct SetGroupByFilterOp
+{
+    typedef typename tree::LeafManager<PointDataTree>   LeafManagerT;
+    typedef typename LeafManagerT::LeafRange            LeafRangeT;
+    typedef typename PointDataTree::LeafNodeType        LeafNodeT;
+    typedef AttributeSet::Descriptor::GroupIndex        GroupIndex;
+    typedef typename FilterT::Data                      FilterDataT;
+
+    SetGroupByFilterOp( const GroupIndex& index, const FilterDataT& filterData)
+        : mIndex(index)
+        , mFilterData(filterData) { }
+
+    void operator()(const typename LeafManagerT::LeafRange& range) const
+    {
+        for (typename LeafManagerT::LeafRange::Iterator leaf=range.begin(); leaf; ++leaf) {
+
+            // obtain the group attribute array
+
+            GroupWriteHandle group(leaf->groupWriteHandle(mIndex));
+
+            // create the filter
+
+            FilterT filter(FilterT::create(*leaf, mFilterData));
+
+            // Use an IndexOnIter if a voxel Coord is required, otherwise an IndexIter will be faster
+
+            if (FilterTraits<FilterT>::RequiresCoord) {
+                typename LeafNodeT::IndexOnIter iter = leaf->beginIndexOn();
+                FilterIndexIter<typename LeafNodeT::IndexOnIter, FilterT> filterIndexIter(iter, filter);
+
+                for (; filterIndexIter; ++filterIndexIter) {
+                    group.set(*filterIndexIter, true);
+                }
+            }
+            else {
+                IndexIter iter = leaf->beginIndexAll();
+                FilterIndexIter<IndexIter, FilterT> filterIndexIter(iter, filter);
+
+                for (; filterIndexIter; ++filterIndexIter) {
+                    group.set(*filterIndexIter, true);
+                }
+            }
```

Comment:
Hmm I've taken another look and seen that this becomes slightly more tricky because the FilterIndexIter is templated on the indexIter type and its impossible to get the coord with a generic IndexIter. I was hoping to do something like this:

```
template<typename LeafNodeT, typename IndexIterT>
struct IndexIteratorTraits {};

template<typename LeafNodeT>
struct IndexIteratorTraits<LeafNodeT, typename LeafNodeT::IndexIter>
{
    typedef typename LeafNodeT::IndexIter IterT;
    static const bool HasCoord = false;
    static IterT begin(LeafNodeT& leaf) { return leaf.beginIndexAll(); }
};

template<typename LeafNodeT>
struct IndexIteratorTraits<LeafNodeT, typename LeafNodeT::IndexOnIter>
{
    typedef typename LeafNodeT::IndexOnIter IterT;
    static const bool HasCoord = true;
    static IterT begin(LeafNodeT& leaf) { return leaf.beginIndexOn(); }
};

template<typename LeafNodeT>
struct IndexIteratorTraits<LeafNodeT, typename LeafNodeT::IndexOffIter>
{
    typedef typename LeafNodeT::IndexOffIter IterT;
    static const bool HasCoord = true;
    static IterT begin(LeafNodeT& leaf) { return leaf.beginIndexOff(); }
};

template <typename PointDataTree, typename FilterT, typename IndexIterT = tools::IndexIter>
struct SetGroupByFilterOp
{
    typedef typename PointDataTree::LeafNodeType   LeafNodeT;

    void operator()(const typename LeafManagerT::LeafRange& range) const
    {
        for (typename LeafManagerT::LeafRange::Iterator leaf=range.begin(); leaf; ++leaf) 
        {
            ...

            if (FilterTraits<FilterT>::RequiresCoord && 
                IndexIteratorTraits<IndexIterT>::HasCoord) 
            {
                IndexIterT iter = IndexIteratorTraits<LeafNodeT, IndexIterT>(*leaf)::begin();
                ...
            }
            else if (FilterTraits<FilterT>::RequiresCoord && 
                     !IndexIteratorTraits<IndexIterT>::HasCoord) 
            {
                ...
            }
       }
   }
};

```
Wanted to add a static Coord method on the traits but this still suffers from the same limitation. I'm not sure the best solution here.